### DISCUSSION TITLE
SSL_CTX_set_keylog_callback not invoked for handshake failure scenario for TLS1.3

### DISCUSSION THREAD

**Discussion by PrakashGibbs** [openssl/openssl#25721] [Q&A] [answered]
Hi Team,

I have set SSL_CTX_set_keylog_callback(ctx, KeyLogCBFunction); to get the Keys in each step of my TLS 1.3 handshake and finally to decrypt the pcap using wireshark so far.
Unfortunately, in one of the scenarios, i'm getting the bellow error after client hello and server hello.

```
callback:3568 client(fd=0x400):   >>> TLS 1.3(0x304)  Alert [length 0002], fatal illegal_parameter  RA:0x7ffff6afeae8
callback:3204 SSL alert write(fd=0x400):fatal:illegal parameter
callback:3215 SSL_connect(fd=0x400): -1 return value during error  RA:0x7ffff6b36691
handshake:143 SSL_connect returned ret=-1
handshake:201 SSL_connect error (-1) for fd=0x400, ssl_err=1, errno=0()
handshake:211 SSL_connect error (-1) for fd=0x400, ssl_err=1, errno=0()
ssl_internal_errors:29 handshake
map_error_code:3172 ppm_tls_ctr_map_error_code: reason:999 ctr=195
ssl_internal_errors:59 SSL error 999 [error:0A0003E7:SSL routines::invalid session id] ssl/statem/statem_clnt.c:tls_process_server_hello:1490:
```
![image](https://github.com/user-attachments/assets/0d46ae6c-2e89-4278-861d-5c0500423925)

So, i wanted to debug further by decrypting the Handshake message which are encrypted to get the parameters that my client is sending to server but the callback function set using SSL_CTX_set_keylog_callback is not called after client/server hello post which i get the illegal parameter error.

Could you guys clarify me will the keylog callback function not be called to atleast give me the secrets available till server hello?
How can i decrypt the server hello Application Data and the next packet which is also encrypted?

Please excuse if any of my understanding is wrong.
Thanks,

**Comment by mattcaswell** [comment, marked as answer]
The error message can tell us quite a lot here. It appears that the client is aborting the handshake complaining that the server has sent it an illegal parameter while processing the ServerHello. The problem appears to be in the `legacy_session_id_echo` field of the ServerHello. The TLSv1.3 RFC says this about the field:

````
   legacy_session_id_echo:  The contents of the client's
      legacy_session_id field.  Note that this field is echoed even if
      the client's value corresponded to a cached pre-TLS 1.3 session
      which the server has chosen not to resume.  A client which
      receives a legacy_session_id_echo field that does not match what
      it sent in the ClientHello MUST abort the handshake with an
      "illegal_parameter" alert.
````

i.e. The server is expected to echo back the legacy_session_id that the client sent to it in the ClientHello. For some reason this does not seem to be correct here which is a protocol violation by the server. In wireshark check the value of the ClientHello.legacy_session_id field and compare it to the value in the ServerHello.legacy_session_id_echo field. They should be the same.

You aren't getting the callback invoked because the problem occurs before the client gets as far as establishing the keys to decrypt the rest of the server messages.

**Reply by PrakashGibbs** [reply]
Thanks for the detailed response @mattcaswell. 
I see session_id field but not he legacy_session_id in both client and server hello.
![Screenshot 2024-10-17 at 10 01 33 PM](https://github.com/user-attachments/assets/7c838bbc-9602-4cfc-8010-ee095cf148c9)

As shown in first the screenshot, the server hello's application data and the next packet from the server are encrypted, any idea if that can be decrypted just to see what server has sent to us to debug the illegal parameter values details in the wireshark as the logs are sufficient in our case? Please throw some light on this. Thanks

**Reply by mattcaswell** [reply]
> I see session_id field but not he legacy_session_id in both client and server hello.

It's the same thing. In previous versions of TLS (i.e. before TLSv1.3) it was just called "session_id". This is a TLSv1.3 handshake, but apparently Wireshark is using this older terminology.

> As shown in first the screenshot, the server hello's application data and the next packet from the server are encrypted, any idea if that can be decrypted just to see what server has sent to us to debug the illegal parameter values details

The ServerHello is not and never is encrypted. The "Appication Data" that you see in the first Wireshark screenshoot are the encrypted subsequent messages in the first server flight (e.g EncryptedExtensions, Certificate, CertificateVerify, Finished). It won't be easy to decrypt those messages because the keys have not been generated yet by OpenSSL. But you don't need to. The problem is not with those messages. OpenSSL is emitting the illegal_parameter alert before it gets as far as looking at those.

Like I said above it appears that OpenSSL does not like the session id in the ServerHello. Please show the value for session id in both the ClientHello and the ServerHello. They should be identical.

Also, please confirm which exact version of OpenSSL you are using.

**Reply by PrakashGibbs** [reply]
Server hello had empty session ID

![image](https://github.com/user-attachments/assets/f2cc410d-e58c-436d-b6ed-d56eefe27ce9)

If Client Hello has any valid session ID then the server hello has to echo back the same ? is that the reason for this error.?
Also i'm looking for an RFC that mandates this, in case if you have it handy please share that would be helpful. Many thanks.

**Reply by mattcaswell** [reply]
> If Client Hello has any valid session ID then the server hello has to echo back the same ?

Yes

> is that the reason for this error.?

Yes

> Also i'm looking for an RFC that mandates this, in case if you have it handy please share that would be helpful

See the quote from the TLSv1.3 RFC that I gave in my first [response](https://github.com/openssl/openssl/discussions/25721#discussioncomment-10968410) to this thread.

That comes from [section 4.1.3 in RFC 8446](https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.3) - in particular the description of the "legacy_session_id_echo" field (which Wireshark is just describing as "Session ID")

What can you tell me about this server? Do you operate it? It appears to be non-compliant with the spec.

**Reply by mattcaswell** [reply]
A workaround to this problem might be to turn off middlebox compatibility mode in OpenSSL (which is on by default). One of the effects of turning this mode off is that it causes the ClientHello to contain a 0 length legacy_session_id:

```` C
    SSL_CTX_clear_options(ctx, SSL_OP_ENABLE_MIDDLEBOX_COMPAT);
````

https://docs.openssl.org/master/man3/SSL_CTX_set_options/#notes

The disadvantage of turning this off is that some old middleboxes in between the client and server might start dropping connections spuriously.

**Comment by PrakashGibbs** [comment]
Thanks very much @mattcaswell  really appreciate.
It was one of the internal tool. may be i can check with the guy who operates it and make it compliant with RFC.
